# Task 3 - Loan Approval Prediction

## Data Exploration

In [1]:
import pandas as pd

In [2]:
loan_data = pd.read_csv("loan_approval_dataset.csv")

In [3]:
loan_data.head()

loan_id   no_of_dependents      education  self_employed   income_annum  \
0        1                  2       Graduate             No        9600000   
1        2                  0   Not Graduate            Yes        4100000   
2        3                  3       Graduate             No        9100000   
3        4                  3       Graduate             No        8200000   
4        5                  5   Not Graduate            Yes        9800000   

    loan_amount   loan_term   cibil_score   residential_assets_value  \
0      29900000          12           778                    2400000   
1      12200000           8           417                    2700000   
2      29700000          20           506                    7100000   
3      30700000           8           467                   18200000   
4      24200000          20           382                   12400000   

    commercial_assets_value   luxury_assets_value   bank_asset_value  \
0                  17600000              22700000            8000000   
1                   2200000               8800000            3300000   
2                   4500000              33300000           12800000   
3                   3300000              23300000            7900000   
4                   8200000              29400000            5000000   

   loan_status  
0     Approved  
1     Rejected  
2     Rejected  
3     Rejected  
4     Rejected

In [4]:
loan_data.tail()

loan_id   no_of_dependents      education  self_employed   income_annum  \
4264     4265                  5       Graduate            Yes        1000000   
4265     4266                  0   Not Graduate            Yes        3300000   
4266     4267                  2   Not Graduate             No        6500000   
4267     4268                  1   Not Graduate             No        4100000   
4268     4269                  1       Graduate             No        9200000   

       loan_amount   loan_term   cibil_score   residential_assets_value  \
4264       2300000          12           317                    2800000   
4265      11300000          20           559                    4200000   
4266      23900000          18           457                    1200000   
4267      12800000           8           780                    8200000   
4268      29700000          10           607                   17800000   

       commercial_assets_value   luxury_assets_value   bank_asset_value  \
4264                    500000               3300000             800000   
4265                   2900000              11000000            1900000   
4266                  12400000              18100000            7300000   
4267                    700000              14100000            5800000   
4268                  11800000              35700000           12000000   

      loan_status  
4264     Rejected  
4265     Approved  
4266     Rejected  
4267     Approved  
4268     Approved

In [5]:
loan_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4269 entries, 0 to 4268
Data columns (total 13 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   loan_id                    4269 non-null   int64 
 1    no_of_dependents          4269 non-null   int64 
 2    education                 4269 non-null   object
 3    self_employed             4269 non-null   object
 4    income_annum              4269 non-null   int64 
 5    loan_amount               4269 non-null   int64 
 6    loan_term                 4269 non-null   int64 
 7    cibil_score               4269 non-null   int64 
 8    residential_assets_value  4269 non-null   int64 
 9    commercial_assets_value   4269 non-null   int64 
 10   luxury_assets_value       4269 non-null   int64 
 11   bank_asset_value          4269 non-null   int64 
 12   loan_status               4269 non-null   object
dtypes: int64(10), object(3)
memory usage: 433.7+ KB


In [6]:
loan_data.isnull().sum()

loan_id                      0
 no_of_dependents            0
 education                   0
 self_employed               0
 income_annum                0
 loan_amount                 0
 loan_term                   0
 cibil_score                 0
 residential_assets_value    0
 commercial_assets_value     0
 luxury_assets_value         0
 bank_asset_value            0
 loan_status                 0
dtype: int64

In [7]:
loan_data.describe()

loan_id   no_of_dependents   income_annum   loan_amount  \
count  4269.000000        4269.000000   4.269000e+03  4.269000e+03   
mean   2135.000000           2.498712   5.059124e+06  1.513345e+07   
std    1232.498479           1.695910   2.806840e+06  9.043363e+06   
min       1.000000           0.000000   2.000000e+05  3.000000e+05   
25%    1068.000000           1.000000   2.700000e+06  7.700000e+06   
50%    2135.000000           3.000000   5.100000e+06  1.450000e+07   
75%    3202.000000           4.000000   7.500000e+06  2.150000e+07   
max    4269.000000           5.000000   9.900000e+06  3.950000e+07   

         loan_term   cibil_score   residential_assets_value  \
count  4269.000000   4269.000000               4.269000e+03   
mean     10.900445    599.936051               7.472617e+06   
std       5.709187    172.430401               6.503637e+06   
min       2.000000    300.000000              -1.000000e+05   
25%       6.000000    453.000000               2.200000e+06   
50%      10.000000    600.000000               5.600000e+06   
75%      16.000000    748.000000               1.130000e+07   
max      20.000000    900.000000               2.910000e+07   

        commercial_assets_value   luxury_assets_value   bank_asset_value  
count              4.269000e+03          4.269000e+03       4.269000e+03  
mean               4.973155e+06          1.512631e+07       4.976692e+06  
std                4.388966e+06          9.103754e+06       3.250185e+06  
min                0.000000e+00          3.000000e+05       0.000000e+00  
25%                1.300000e+06          7.500000e+06       2.300000e+06  
50%                3.700000e+06          1.460000e+07       4.600000e+06  
75%                7.600000e+06          2.170000e+07       7.100000e+06  
max                1.940000e+07          3.920000e+07       1.470000e+07

# Data Preprocessing

In [8]:
loan_data = loan_data.drop("loan_id" , axis = 1)

In [9]:
loan_data.select_dtypes(['object'])

education self_employed loan_status
0          Graduate            No    Approved
1      Not Graduate           Yes    Rejected
2          Graduate            No    Rejected
3          Graduate            No    Rejected
4      Not Graduate           Yes    Rejected
...             ...           ...         ...
4264       Graduate           Yes    Rejected
4265   Not Graduate           Yes    Approved
4266   Not Graduate            No    Rejected
4267   Not Graduate            No    Approved
4268       Graduate            No    Approved

[4269 rows x 3 columns]

## Encode categorical columns

In [10]:
from sklearn.preprocessing import LabelEncoder
encode = LabelEncoder()
for col in loan_data.select_dtypes(include = "object").columns:
    loan_data[col] = encode.fit_transform(loan_data[col])

In [11]:
loan_data.head()

no_of_dependents  education  self_employed  income_annum  loan_amount  \
0                 2          0              0       9600000     29900000   
1                 0          1              1       4100000     12200000   
2                 3          0              0       9100000     29700000   
3                 3          0              0       8200000     30700000   
4                 5          1              1       9800000     24200000   

   loan_term  cibil_score  residential_assets_value  commercial_assets_value  \
0         12          778                   2400000                 17600000   
1          8          417                   2700000                  2200000   
2         20          506                   7100000                  4500000   
3          8          467                  18200000                  3300000   
4         20          382                  12400000                  8200000   

   luxury_assets_value  bank_asset_value  loan_status  
0             22700000           8000000            0  
1              8800000           3300000            1  
2             33300000          12800000            1  
3             23300000           7900000            1  
4             29400000           5000000            1

## Check class Imbalance

In [12]:
loan_data.columns

Index([' no_of_dependents', ' education', ' self_employed', ' income_annum',
       ' loan_amount', ' loan_term', ' cibil_score',
       ' residential_assets_value', ' commercial_assets_value',
       ' luxury_assets_value', ' bank_asset_value', ' loan_status'],
      dtype='object')

In [13]:
loan_data[" loan_status"].value_counts()

 loan_status
0    2656
1    1613
Name: count, dtype: int64

# Data Split into Train and Test

In [14]:
x = loan_data.drop([" loan_status"] , axis = 1 )
x

no_of_dependents  education  self_employed  income_annum  loan_amount  \
0                    2          0              0       9600000     29900000   
1                    0          1              1       4100000     12200000   
2                    3          0              0       9100000     29700000   
3                    3          0              0       8200000     30700000   
4                    5          1              1       9800000     24200000   
...                ...        ...            ...           ...          ...   
4264                 5          0              1       1000000      2300000   
4265                 0          1              1       3300000     11300000   
4266                 2          1              0       6500000     23900000   
4267                 1          1              0       4100000     12800000   
4268                 1          0              0       9200000     29700000   

      loan_term  cibil_score  residential_assets_value  \
0            12          778                   2400000   
1             8          417                   2700000   
2            20          506                   7100000   
3             8          467                  18200000   
4            20          382                  12400000   
...         ...          ...                       ...   
4264         12          317                   2800000   
4265         20          559                   4200000   
4266         18          457                   1200000   
4267          8          780                   8200000   
4268         10          607                  17800000   

      commercial_assets_value  luxury_assets_value  bank_asset_value  
0                    17600000             22700000           8000000  
1                     2200000              8800000           3300000  
2                     4500000             33300000          12800000  
3                     3300000             23300000           7900000  
4                     8200000             29400000           5000000  
...                       ...                  ...               ...  
4264                   500000              3300000            800000  
4265                  2900000             11000000           1900000  
4266                 12400000             18100000           7300000  
4267                   700000             14100000           5800000  
4268                 11800000             35700000          12000000  

[4269 rows x 11 columns]

In [15]:
y = loan_data[" loan_status"]
y

0       0
1       1
2       1
3       1
4       1
       ..
4264    1
4265    0
4266    1
4267    0
4268    0
Name:  loan_status, Length: 4269, dtype: int32

In [16]:
from sklearn.model_selection import train_test_split

In [17]:
x_train , x_test, y_train,y_test = train_test_split(x , y , test_size = 0.3 , random_state = 42)

In [18]:
from sklearn.tree import DecisionTreeClassifier
tree_model = DecisionTreeClassifier(criterion = "entropy", max_depth = 5 , min_samples_split = 10 , random_state = 42, )
tree_model.fit(x_train , y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=5, min_samples_split=10,
                       random_state=42)

In [19]:
y_pred = tree_model.predict(x_test)

# Evaluate Model

In [20]:
from sklearn.metrics import classification_report

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.96      0.97       810
           1       0.94      0.98      0.96       471

    accuracy                           0.97      1281
   macro avg       0.96      0.97      0.97      1281
weighted avg       0.97      0.97      0.97      1281



# SMOTE to Address class Imbalance 

In [21]:
from imblearn.over_sampling import SMOTE

smote = SMOTE(random_state = 42)
x_train_smote , y_train_smote = smote.fit_resample(x_train,y_train)

print(f"Before Smote: {y_train.value_counts()}")
print(f"After Smote: {y_train_smote.value_counts()}")

Before Smote:  loan_status
0    1846
1    1142
Name: count, dtype: int64
After Smote:  loan_status
1    1846
0    1846
Name: count, dtype: int64


# Train Logistic Regression On Balance Data

In [22]:
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression(max_iter = 1000 , random_state = 42)
lr_model.fit(x_train_smote , y_train_smote)

LogisticRegression(max_iter=1000, random_state=42)

In [23]:
y_pred_smote = lr_model.predict(x_test)

In [24]:
print(classification_report(y_test, y_pred_smote))

              precision    recall  f1-score   support

           0       0.84      0.83      0.83       810
           1       0.71      0.73      0.72       471

    accuracy                           0.79      1281
   macro avg       0.78      0.78      0.78      1281
weighted avg       0.79      0.79      0.79      1281



# Bonus Experiment – SMOTE + Logistic Regression vs. Decision Tree
**A Decision Tree classifier trained on the original imbalanced dataset achieved very high accuracy (97%) with strong precision and recall for both classes, indicating it handled the imbalance well.
When SMOTE was applied to balance the training data and Logistic Regression was used, performance on the original test set dropped to 79% accuracy, with noticeably lower precision and recall. This suggests that in this dataset, synthetic oversampling with a linear model introduced noise and reduced generalization, whereas the non-linear Decision Tree was more capable of capturing the patterns without balancing.**